### PR TITLE
Nickname call response

### LINKED ISSUE #123: Beep bops
Get the bot to reply with beep boops when you say his name

### PR DESCRIPTION
**Pull request by RaoulHC**
Adds an option to set a response in features, which will be sent whenever a user says url-bot-rs's nick.

Based on #158.

Closes issue #123

### CODE REVIEW

**Comment by nuxeh** [comment]
I've tested this (the call-response changes) and have noticed a number of potential functional issues:
 - Messages containing multiple instances of the bot user name result in multiple responses, e.g.
```
22:03 <@ed> wall-e wall-e wall-e wall-e
22:03 < wall-e> beep bop beep bop 
22:03 < wall-e> beep bop beep bop 
22:03 < wall-e> beep bop beep bop 
22:04 < wall-e> beep bop beep bop
```
- Pings don't work, so `wall-e: hello!` wouldn't do anything.
- Messages containing the nick and a URL will produce both a call-response and a URL response, e.g.
```
22:04 <@ed> wall-e google.com
22:04 < wall-e> beep bop beep bop
22:04 < wall-e> ⤷ Google
```

And a couple of other small things...

- It's nice that the response is exposed in the configuration, however:
  - Since `response` has type `Option<String>`, the default (if it's not explicitly given a value in the configuration) is `None`, and is therefore not present in a written-out default configuration. This seems like a missed opportunity for self-documentation purposes.
  - As the name for the configuration option, `response` isn't immediately apparent or obvious what it does. Maybe clearer to have something like `nick_response` instead?
   - Perhaps also clearer to have e.g. `nick_response` under `[features]`, and an accompanying parameter in `[parameters]`?
- There are some formatting and style issues that I would like to discuss before merging this, along with the above functional issues.

**Comment by nuxeh** [comment]
Also, while I like the idea of splitting out a `get_title` function, in this PR, that function becomes one with some strange side effects... Mainly adding entries to the database, which I don't think is a particularly clean design.

Personally I'd prefer to keep the logic in the `privmsg` function for now. This might also work reasonably well, together with the kind of logic that would be needed to refactor according to some of the functional issues mentioned above.

**Comment by nuxeh** [comment]
And, thanks for the PR @RaoulHC :)

**Comment by RaoulHC** [comment]
Gone for a different approach, leaving the logic in `privmsg`, and resulting in a much smaller change.

I think this resolves all of your issues, resolving pings in the form `wall-e: foo` and `wall-e, bar`, as well as only responding once, if the nick was found and there are no url in the message. I've also changed the configuration to be a `nick_response` boolean in features and `nick_response_str` in parameters, though I'm not 100% sure on that name.

Let me know what you think :)

**Comment by nuxeh** [comment]
Hi @RaoulHC, nice, I like the direction this is going in, however i raise you an even simpler solution...

Before the per-token loop you could have something like:

```rust
let nick_seen = msg.contains(client.current_nickname());
```

And at the end of the function:

```rust
    let nick_resp = rtd.conf.features.nick_response;
    if nick_resp && nick_seen && num_processed == 0 {
        client.send_privmsg(target, &rtd.conf.params.nick_response).unwrap();
    }
```

Which, as far as I can tell, achieves the same thing. (?)

A few notes:
- using [current_nickname()](https://docs.rs/irc/0.13.6/irc/client/struct.IrcClient.html#method.current_nickname) gives the actual runtime nickname, rather than the configured one, which could be different, e.g. appended with `_`.
- `nick_response_str` could simply be called `params.nick_response`
- A final touch would be to have a note in the readme about the new feature.

File: src/message.rs [hunk 1]
```diff
@@ -78,13 +78,27 @@ fn privmsg(client: &IrcClient, message: &Message, rtd: &Rtd, db: &Database, targ
     let mut num_processed = 0;
     let mut dedup_urls = HashSet::new();
 
+    // flags to mark whether we've got a ping or url
+    let mut nick_ping: bool = false;
```
Comment (nuxeh) [inline src/message.rs]: Fwiw, you can leave out the type annotation here.

**Comment by nuxeh** [comment]
Although another thing is that, if URLs are present, but all of them error, we will get a beep boop...

**Comment by nuxeh** [comment]
In which case, could keep a counter of errors from `resolve_url` and only respond when that's also 0, perhaps.

**Comment by RaoulHC** [comment]
> Which, as far as I can tell, achieves the same thing. (?)

With `msg.contains` you would have it respond if the name appears in the middle of another word, which for short names could be quite common so I'm not sure this approach is ideal.

**Comment by nuxeh** [comment]
> With `msg.contains` you would have it respond if the name appears in the middle of another word, which for short names could be quite common so I'm not sure this approach is ideal.

Yes, very true

**Pull request closed without merging**